arangod/RestHandler/RestTestHandler.cpp

```diff
@@ -0,0 +1,163 @@
+/// @brief was docuBlock JSF_batch_processing
+////////////////////////////////////////////////////////////////////////////////
+
+ResultT<RequestLane> RestTestHandler::requestLaneFromString(const std::string &str)
+{
+  #define LANE_ENTRY(s) {#s, RequestLane:: s},
+  static std::map<std::string, RequestLane> lanes = {
+    LANE_ENTRY(CLIENT_FAST)
+    LANE_ENTRY(CLIENT_AQL)
+    LANE_ENTRY(CLIENT_V8)
+    LANE_ENTRY(CLIENT_SLOW)
+    LANE_ENTRY(AGENCY_INTERNAL)
+    LANE_ENTRY(AGENCY_CLUSTER)
+    LANE_ENTRY(CLUSTER_INTERNAL)
+    LANE_ENTRY(CLUSTER_V8)
+    LANE_ENTRY(CLUSTER_ADMIN)
+    LANE_ENTRY(SERVER_REPLICATION)
+    LANE_ENTRY(TASK_V8)
+  };
+
+  auto entry = lanes.find(str);
+
+  if (entry != lanes.end()) {
+    return entry->second;
+  }
+
+  return Result(TRI_ERROR_HTTP_BAD_PARAMETER,
+    "Expected request-lane, found `" + str + "`");
+}
+
+
+RestStatus RestTestHandler::execute() {
+  // extract the request type
+  auto const type = _request->requestType();
+
+  if (type != rest::RequestType::POST) {
+    generateError(rest::ResponseCode::METHOD_NOT_ALLOWED,
+                  TRI_ERROR_HTTP_METHOD_NOT_ALLOWED);
+    return RestStatus::DONE;
+  }
+
+  // Get the execution class
+  // then optionally get the duration to work
+  // then check if we should be a waiting rest handler
+
+  auto const &suffixes = _request->suffixes();
+
+  if (suffixes.size() != 1) {
+    generateError(rest::ResponseCode::NOT_FOUND,
+            TRI_ERROR_HTTP_NOT_FOUND,
+            "expecting GET /_devel/test/<request-lane>");
+    return RestStatus::DONE;
+  }
+
+  auto res = requestLaneFromString(suffixes[0]);
+
+  if (res.fail()) {
+    generateError(res);
+    return RestStatus::DONE;
+  }
+
+  bool parsingSuccess = false;
+  VPackSlice const body = this->parseVPackBody(parsingSuccess);
+  if (!parsingSuccess) {
+    return RestStatus::DONE;
+  }
+
+  if (!body.isObject()) {
+    generateError(rest::ResponseCode::BAD, TRI_ERROR_TYPE_ERROR,
+                  "expecting JSON object body");
+    return RestStatus::DONE;
+  }
+
+  LOG_TOPIC(TRACE, Logger::FIXME) << "Generating work on lane " << suffixes[0];
+
+  clock::duration duration = std::chrono::milliseconds(100);
+
+  if (body.hasKey("workload")) {
+
+    auto workload = body.get("workload");
+
+    if (workload.isNumber()) {
+      duration = std::chrono::milliseconds(workload.getInt());
+    } else {
+      generateError(rest::ResponseCode::BAD, TRI_ERROR_TYPE_ERROR,
+                  "expecting int for `workload`");
+      return RestStatus::DONE;
+    }
+  }
+
+  auto self(shared_from_this());
+
+  bool ok = SchedulerFeature::SCHEDULER->queue(
+    PriorityRequestLane(res.get()),
+    [this, self, duration]() {
+      auto stop = clock::now() + duration;
+
+      // Please think of a better method to generate work.
+      //  Do we actually need work or is a sleep ok?
+      while (clock::now() < stop) {
+        for (int i = 0; i < 10000; i++) {
+          _count += i * i;
```

Comment:
Not sure what `_count` is for, it's probably a dummy. But it is never initialized, so at least ASAN will complain about it.

Reply:
It's used to generate work load. It is a protected member, so that the compiler can not guess that the result is never used.
Initialization fixed.

Reply:
The compiler will be able to guess it once we enable LTO and whole program optimization :-)

Reply:
Argument accepted!